---
title: "00-import"
date: "Compiled at `r format(Sys.time(), '%Y-%m-%d %H:%M:%S', tz = 'UTC')` UTC"
output: github_document
params:
  name: "00-import" # change if you rename file
---

```{r here, message=FALSE}
here::i_am(paste0(params$name, ".Rmd"), uuid = "a0e8b3cd-3c98-4629-928f-f573d12fd0f4")
```

The purpose of this document is ...

```{r packages}
library("conflicted")
library(here)
library(tidyverse)
library(data.table)
library(TreeSummarizedExperiment)
library(mia)
library(miaViz)
library(scater)
library(patchwork)
library(tidyverse)
library(viridis)
library(hrbrthemes)
library(circlize)
library(networkD3)
library(pheatmap)

prj_path <- normalizePath("..")
dpath <- here(prj_path, "pipeline/report")

source(here(prj_path, "R/utils.R"))
```

```{r directories}
# create or *empty* the target directory, used to write this file's data: 
projthis::proj_create_dir_target(params$name, clean = F)

# function to get path to target directory: path_target("sample.csv")
path_target <- projthis::proj_path_target(params$name)

# function to get path to previous data: path_source("00-import", "sample.csv")
path_source <- projthis::proj_path_source(params$name)
```

## Tasks

```{r}
fp_metadata <- here(prj_path, "data/metadata.xlsx")
fp_vir_taxa <- here(dpath, "taxonomy.tsv")
# fp_vir_counts <- here(dpath, "abundance_contigs_count.tsv.gz")
# fp_vir_trmean <- here(dpath, "abundance_contigs_trimmed_mean.tsv.gz")
# fp_vir_covfrac <- here(dpath, "abundance_contigs_covered_fraction.tsv.gz")
# fp_vir_tpm <- here(dpath, "abundance_contigs_tpm.tsv.gz")
fp_vir_counts <- here(dpath, "sid90/abundance_contigs_count.tsv.gz")
fp_vir_trmean <- here(dpath, "sid90/abundance_contigs_trimmed_mean.tsv.gz")
fp_vir_covfrac <- here(dpath, "sid90/abundance_contigs_covered_fraction.tsv.gz")
fp_checkv <- here(dpath, "quality_summary.tsv")
fp_virsorter2 <- here(dpath, "final-viral-score.tsv")
fp_vcontact2 <- here(dpath, "genome_by_genome_overview.csv")
fp_lifestyle <- here(dpath, "scaffolds.fna.bacphlip")
fp_virhost <- here(dpath, "Host_prediction_to_genus_m90.csv")
fp_dram_stats <- here(dpath, "vMAG_stats.tsv")
fp_dram_anno <- here(dpath, "annotations.tsv")
fp_vgene_abund <- here(dpath, "abundance_orfs.tsv")

smeta <- openxlsx::read.xlsx(fp_metadata) %>% 
  mutate(sample_id = as.factor(sample_id))
df_checkv <- read_CheckV(fp_checkv)
df_virsorter2 <- read_VirSorter2(fp_virsorter2)
vc2 <- read_vConTACT2(fp_vcontact2)
df_vcontact2 <- vc2$vc_tbl
anno_vcontact2_cluster <- vc2$vc_annotated
df_vtaxa <- read_vtaxonomy(fp_vir_taxa)  # will be replaced or removed
df_vrcyc <- read_lifestyle(fp_lifestyle)
# anno_virhost <- fread(fp_virhost)
df_vgene_abund <- fread(fp_vgene_abund)
df_host <- fread(fp_virhost) %>% 
  setnames(colnames(.), c("contig_id", "aai", "lineage", "score", "methods")) %>% 
  dplyr::select(c(contig_id, lineage)) %>% 
  separate(lineage, sep = ";", into = c("domain", "phylum", "class", "order", "family", "genus"))

df_dramstats <- fread(fp_dram_stats) %>% 
  column_to_rownames("V1") %>% 
  setnames(colnames(.), make.names(colnames(.))) %>% 
  dplyr::select(-c(VIRSorter.category)) %>% 
  setnames(colnames(.), paste0("dramv_", colnames(.))) %>% 
  rownames_to_column("virsorter2_contig_id")
```

```{r}
col_taxa <- c("vtaxa_Phylum", "vtaxa_Order", "vtaxa_Class", "vtaxa_Family", "vtaxa_Genus", "vtaxa_Species")
anno_all <- df_checkv %>% 
  left_join(df_vcontact2, by = "Contig") %>% 
  left_join(df_virsorter2, by = "Contig") %>% 
  # left_join(df_lifestyle, by = "virsorter2_contig_id") %>%
  left_join(df_vtaxa, by = "virsorter2_contig_id") %>%
  left_join(df_dramstats, by = "virsorter2_contig_id") %>%
  mutate(checkv_contig_length = as.integer(checkv_contig_length)) %>% 
  mutate(virsorter2_category = as.integer(virsorter2_category)) %>% 
  arrange(desc(checkv_contig_length)) %>% 
  setnames(col_taxa, str_replace(col_taxa, "vtaxa_", ""))


anno_viral <- anno_all %>% 
  dplyr::filter(!is.na(virsorter2_contig_id))
  # dplyr::filter(!checkv_quality %in% c('Not-determined'))
```

```{r}
read_abundance2 <- function(fp, anno_vir) {
  df <- read_abundance(fp, anno_viral) %>% 
    column_to_rownames("Contig")
  return(df)
}

# df_counts <- read_abundance(fp_vir_counts, anno_viral)
df_counts <- read_abundance(fp_vir_counts, anno_viral, smeta$sample_id)
df_trmean <- read_abundance(fp_vir_trmean, anno_viral, smeta$sample_id)
df_covfrac <- read_abundance(fp_vir_covfrac, anno_viral, smeta$sample_id)
df_abundance <- create_viral_abundance(df_trmean, df_covfrac, covfrac_threshold=0.2) %>% 
  t()
df_abundance_rel <- df_abundance / rowSums(df_abundance)
df_abundance_mclr <- SPRING::mclr(df_abundance_rel) %>% t() %>% 
  data.frame() %>% 
  rownames_to_column("Contig")

# -------------
df_counts_cov <- create_viral_abundance(df_counts, df_covfrac, covfrac_threshold=0.7)
df_trmean_cov <- create_viral_abundance(df_trmean, df_covfrac, covfrac_threshold=0.7)
ctg_len <- data.frame(Contig=rownames(df_counts_cov)) %>% 
  left_join(anno_viral) %>% 
  dplyr::select(c(Contig, checkv_contig_length)) %>% 
  pull(checkv_contig_length)

counts2rpkm <- function(df_counts, seq_len) {
  # For each sample (column)
  # rpkm = num_reads_in_region / (total_reads / 10^6) / (region_len / 10^3)
  #      = num_reads_in_region * 10^9 / total_reads / region_len
  libsize = colSums(df_counts)   #  lib_size: total_reads
  counts_norm_by_libsize = sweep(df_counts * 10^6, 2, libsize, FUN = '/')
  df_rpkm <- sweep(counts_norm_by_libsize * 10^3, 1, seq_len, FUN = '/')
  return(df_rpkm)
}

counts2tpm <- function(df_counts, seq_len, ceiling=FALSE) {
  # For each sample (column)
  # norm_by_len = num_reads_in_region / region_len
  # libsize = sum(norm_by_len)
  # tpm = norm_by_len / (libsize / 10^6)
  #     = norm_by_len * 10^6 / libsize
  norm_by_len = sweep(df_counts, 1, seq_len, FUN = '/')
  libsize = colSums(norm_by_len)
  df_tpm <- sweep(norm_by_len, 2, libsize, FUN = '/') * 10^6
  if (ceiling) {
    df_tpm <- ceiling(df_tpm)
  }
  return(df_tpm)
}

mx_rpkm <- counts2rpkm(df_counts_cov, ctg_len)
mx_tpm <- counts2tpm(df_counts_cov, ctg_len, TRUE)
mx_tmm <- sweep(df_trmean_cov, 2, colSums(df_trmean_cov), FUN = '/') * 10^6
mx_tmm <- ceiling(mx_tmm)
```


```{r TSE}
abundance_matrix <- function(df) {
  rst <- df %>% 
    column_to_rownames("Contig") %>% 
    as.matrix()
  return(rst)
}

feature_anno <- anno_viral %>% 
  column_to_rownames("Contig") %>%
  t() %>% 
  data.frame() %>% 
  dplyr::select(rownames(mx_tpm)) %>% 
  t() %>% 
  DataFrame()

tbl_counts <- abundance_matrix(df_counts)
tbl_trmean <- abundance_matrix(df_trmean)
tbl_covfrac <- abundance_matrix(df_covfrac)

tse <- TreeSummarizedExperiment(assays = SimpleList(counts = as.matrix(ceiling(tbl_counts))),
                                colData = smeta,
                                rowData = feature_anno)

assay(tse, "rpkm") <- edgeR::rpkm(assay(tse, "counts"), ctg_len)
assay(tse, "tpm") <- counts2tpm(assay(tse, "counts"), ctg_len, TRUE)

tse_vc <- mia::mergeRows(tse, rowData(tse)$vConTACT_VC2)
tse_condition <- mia::mergeCols(tse_vc, colData(tse)$condition)
```

```{r}
tse1 <- tse_vc
tse1 <- transformCounts(tse1, method = "relabundance", abund_values = "counts")
tse1 <- transformCounts(tse1, method = "log10", abund_values = "relabundance", pseudocount = 1)


idx_richness <- c("chao1", "ace")
tse1 <- estimateRichness(tse1, abund_values = "counts",
                        index = idx_richness,
                        name = paste0("richness_", idx_richness))

idx_diversity <- c("coverage", "fisher", "shannon")
tse1 <- estimateDiversity(tse1, abund_values = "counts",
                        index = idx_diversity,
                        name = paste0("diversity_", idx_diversity))

idx_evenness <- c("pielou", "camargo", "simpson_evenness", "evar", "bulla")
tse1 <- estimateEvenness(tse1, abund_values = "counts",
                        index = idx_evenness,
                        name = paste0("evenness_", idx_evenness))
plotColData(tse1, "diversity_shannon", "condition")
plotColData(tse1, "richness_chao1", "condition")
plotColData(tse1, "evenness_pielou", "condition")
```

```{r beta diversity}
plot_beta_diversity <- function(tse_obj, name, NMDS=FALSE, scale = F, ...) {
  if (NMDS == TRUE) {
    tse_obj <- runNMDS(tse_obj, FUN = vegan::vegdist, name = name, ...)
    xlab <- "PC 1"
    ylab <- "PC 2"
  } else {
    tse_obj <- runMDS(tse_obj, FUN = vegan::vegdist, name = name, ...)
    e <- attr(reducedDim(tse_obj, name), "eig")
    rel_eig <- 100 * e/sum(e[e>0])
    xlab <- paste("PC 1 (", round(rel_eig[[1]], 2), "%)", sep = "")
    ylab <- paste("PC 2 (", round(rel_eig[[2]], 2), "%)", sep = "")
  }
  p <- plotReducedDim(tse_obj, name, colour_by = "condition") +
    xlab(xlab) +
    ylab(ylab) +
    stat_ellipse(geom = "polygon", alpha=0.1, aes(fill=colour_by), linetype=2) +
    ggtitle(name) +
    theme_bw()
  return(p)
}

p1 <- plot_beta_diversity(tse1, name = "PCoA", method = "bray", exprs_values = "log10", ntop = nrow(tse))
p2 <- plot_beta_diversity(tse1, name = "MDS_euclidean", method = "euclidean", exprs_values = "log10", ntop = nrow(tse))
p3 <- plot_beta_diversity(tse1, name = "NMDS_BC", method = "bray", exprs_values = "log10", ntop = nrow(tse), NMDS = TRUE)
p4 <- plot_beta_diversity(tse1, name = "NMDS_euclidean", method = "euclidean", exprs_values = "log10", ntop = nrow(tse1), NMDS = TRUE)
p1 + p2 + p3 + p4 + plot_layout(guides = "collect")
p2 + p4
```

```{r composition, fig.height=6, fig.width=6}
rank_sel <- "Family"
tse_tmp <- tse_condition
tse_tmp <- relAbundanceCounts(tse_tmp, abund_values = "counts")
tse_rank <- agglomerateByRank(tse_tmp, rank = rank_sel, onRankOnly=T, na.rm=F)
# top_taxa <- getTopTaxa(tse_rank,top = 5, abund_values = "relabundance")
# top_taxa <- top_taxa[str_detect(top_taxa, "NA|___", negate = T)]
taxa_renamed <- lapply(rownames(rowData(tse_rank)),
                   function(x){if (str_detect(x, "^NA|^__", negate = T)) {x} else {"Other"}})
# rowData(tse_rank)[[rank_sel]] <- rownames(rowData(tse_rank))
rowData(tse_rank)[[rank_sel]] <- as.character(taxa_renamed)
plotAbundance(tse_rank, abund_values="relabundance", rank = rank_sel,
              order_rank_by="abund", add_x_text=T) +
  labs(x = "Samples", y = "Relative abundance")
ggsave("fig1_family_abundance.pdf", width = 4, height = 5)

colnames(tbl_counts)
rownames(rowData(tse_rank))
as.character(taxa_renamed)
```


```{r diversity, fig.height=4, fig.width=5}
library(ggpubr)
plot_diversity <- function(tse_obj, col_diversity, metadata) {
  label.y <- max(colData(tse_obj)[[col_diversity]]) * 1.4
  p <- ggviolin(data.frame(colData(tse_obj)), 
                x = "condition", y = col_diversity, color = "condition", 
                palette = "jco", 
                width = 0.3, 
                legend="none") +
    geom_boxplot(aes(color=condition), width=0.03, alpha=0.5) +
    geom_point() +
    stat_compare_means(comparisons = list(c(unique(metadata$condition))), label.y = label.y) +
    labs(x = "Condition")
  return(p)
}

cols_diversity <- c("richness_chao1", "richness_ace", "diversity_coverage", "diversity_fisher", "diversity_shannon", "evenness_pielou", "evenness_camargo", "evenness_simpson_evenness", "evenness_evar", "evenness_bulla")

for (col in cols_diversity) {
  p <- plot_diversity(tse1, col, smeta)
  print(p)
}

fig2_1 <- plot_diversity(tse1, "richness_chao1", smeta)
fig2_2 <- plot_diversity(tse1, "diversity_shannon", smeta)
fig2_3 <- plot_diversity(tse1, "evenness_pielou", smeta)

fig2 <- fig2_1 | fig2_2 | fig2_3
ggsave("fig2_diversity.pdf", plot = fig2, width = 9, height = 4)
```
```{r AMG heatmap todo, fig.width=40}
df_vgene_abund <- fread(fp_vgene_abund)
vgene_len <- pull(df_vgene_abund, Length)
vgene_abund <- df_vgene_abund %>% 
  column_to_rownames("Geneid") %>% 
  dplyr::select(-c(Chr, Start, End, Strand, Length)) %>% 
  as.matrix()

vgene_abund_norm <- sweep(vgene_abund, 1, vgene_len, FUN = "/")
vgene_abund_rel <- 100 * sweep(vgene_abund_norm, 2, colSums(vgene_abund_norm), FUN = "/")
df_vgene_rel <- as.data.frame(vgene_abund_rel) %>% 
  rownames_to_column("Geneid")

df_anno <- fread(fp_dram_anno) %>% 
  mutate(Chr = str_replace(scaffold, "-cat_[0-9]", "")) %>% 
  mutate(Geneid = str_replace(V1, "-cat_[0-9]", "")) %>% 
  inner_join(df_vgene_rel, by = "Geneid") %>% 
  mutate(annotation = ifelse(ko_id == "", pfam_hits, kegg_hit)) %>% 
  mutate(annotation = ifelse(annotation == "", viral_hit, annotation)) %>% 
  mutate(annotation = ifelse(annotation == "", Geneid, annotation)) %>% 
  mutate(annotation = str_replace(annotation, ";.*", ""))

df_amg <- df_anno %>% 
  dplyr::filter(str_detect(amg_flags, "M")) %>% 
  dplyr::filter(auxiliary_score %in% c(1,2,3))

df4heatmap <- df_vgene_rel %>% 
  inner_join(df_amg[, c("Geneid", "annotation")]) %>% 
  column_to_rownames("Geneid") %>% 
  group_by(annotation) %>% 
  summarise(across(everything(), sum)) %>% 
  column_to_rownames("annotation") %>% 
  dplyr::filter(rowSums(.)>0.1) %>% 
  as.matrix()

df4heatmap <- df4heatmap %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  mutate(condition = ifelse(str_detect(condition, "H"), "Health", "UC")) %>%
  group_by(condition) %>% 
  summarise(across(everything(), sum)) %>% 
  column_to_rownames("condition") %>% 
  t()

pheatmap(df4heatmap, scale = "row", width = 10, 
         cluster_rows = T,
         cluster_cols = F,
         fontsize = 14,
         filename = "fig3_AMG_heatmap.pdf")
```

```{r sankey, fig.height=8, fig.width=16}
host_relab <- assay(tse_tmp, "relabundance") %>% 
  as.data.frame() %>% 
  rownames_to_column("contig_id") %>% 
  inner_join(df_host, by = "contig_id") %>% 
  dplyr::select(c(class, Health, UC)) %>% 
  group_by(class) %>% 
  summarise(across(everything(), sum)) %>% 
  dplyr::filter(str_detect(class, "__$", negate = T)) %>% 
  mutate(sumrel = Health + UC) %>% 
  dplyr::filter(sumrel > 0.0001) %>% 
  column_to_rownames("class") 

# Libraries
dft <- assay(tse_rank, "relabundance") %>% 
  data.frame()
dft <- host_relab %>% 
  # t() %>% 
  # data.frame() %>% 
  rownames_to_column("source") %>% 
  dplyr::filter(str_detect(source, "^NA|^__", negate = T)) %>% 
  # mutate(source_cat=str_replace(source, "[0-9].*", "")) %>% 
  # dplyr::select(-source) %>% 
  # group_by(source_cat) %>% 
  pivot_longer(cols=c("Health", "UC"), names_to = "target", values_to = "value") %>% 
  mutate(target=str_replace(target, "[0-9].*", ""))

node_all <- data.frame(name=unique(c(dft$source, dft$target)))
nodet <- data.frame(nodename=node_all$name, nodeid=1:length(node_all$name)) %>%
  mutate(nodeid = nodeid - 1) %>% 
  mutate(source = nodename, target=nodename, IDsource=nodeid, IDtarget=nodeid)
dft2 <- dft %>% 
  left_join(nodet[,c('source', 'IDsource')], by = 'source') %>% 
  left_join(nodet[,c('target', 'IDtarget')], by = 'target') %>% 
  data.frame()

sankeyNetwork(Links = dft2, Nodes = node_all,
                     Source = "IDsource", Target = "IDtarget",
                     Value = "value", NodeID = "name", 
                     sinksRight=F, fontSize=13)
```


## Files written

These files have been written to the target directory, ```r paste0("data/", params$name)```:

```{r list-files-target}
df <- anno_viral %>% 
  inner_join(df_abundance_mclr, by = "Contig") %>% 
  column_to_rownames("virsorter2_contig_id") %>% 
  rownames_to_column("Virus")

fwrite(df, file = path_target("virome_rst.tsv"), sep = "\t")
openxlsx::write.xlsx(df, file = path_target("virome_rst.xlsx"))
projthis::proj_dir_info(path_target())
```
